langserver/xcache.go

```diff
@@ -0,0 +1,43 @@
+package langserver
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+
+	"github.com/sourcegraph/go-langserver/pkg/lspext"
+	"github.com/sourcegraph/jsonrpc2"
+)
+
+// cacheGet will do a xcache/get request for key, and unmarshal the result
+// into v on a cache hit. If it is a cache miss, false is returned. If the
+// client is not a XCacheProvider, cacheGet will always return false.
+func (h *LangHandler) cacheGet(ctx context.Context, conn jsonrpc2.JSONRPC2, key string, v interface{}) bool {
+	if !h.init.Capabilities.XCacheProvider {
+		return false
+	}
+
+	var r json.RawMessage
+	err := conn.Call(ctx, "xcache/get", lspext.CacheGetParams{Key: key}, &r)
+	if err != nil {
+		return false
+	}
+	b := []byte(r)
+	if bytes.Equal(b, []byte("null")) {
+		return false
```

Comment:
That would cause an allocation, since []byte (which json.RawMessage is a type-alias to) is mutable and string is immutable.